napari/plugins/__init__.py

```diff
@@ -1,34 +1,54 @@
+from functools import lru_cache
 from pathlib import Path
 
 from npe2 import PluginManager as _PluginManager
 from npe2 import PluginManifest
 from npe2.manifest.package_metadata import PackageMetadata
 
 from ..settings import get_settings
+from . import _npe2
 from ._plugin_manager import NapariPluginManager
 
 __all__ = ["plugin_manager", "menu_item_template"]
 
-_npe2pm = _PluginManager.instance()
-
-# this is a workaround for the fact that briefcase does not seem to
-# include napari's entry_points.txt in the bundled app, so the builtin plugins
-# don't get detected.  So we just register it manually.  This could potentially
-# be removed when we move to a different bundle strategy
-if 'napari' not in _npe2pm._manifests:
-    mf_file = Path(__file__).parent.parent / 'builtins.yaml'
-    mf = PluginManifest.from_file(mf_file)
-    mf.package_metadata = PackageMetadata.for_package('napari')
-    _npe2pm.register(mf)
-
+#: Template to use for namespacing a plugin item in the menu bar
+menu_item_template = '{}: {}'
 # the main plugin manager instance for the `napari` plugin namespace.
 plugin_manager = NapariPluginManager()
-plugin_manager._initialize()
 
-# Disable plugins listed as disabled in settings, or detected in npe2
-_from_npe2 = {m.package_metadata.name for m in _npe2pm._manifests.values()}
-_toblock = get_settings().plugins.disabled_plugins.union(_from_npe2)
-plugin_manager._blocked.update(_toblock)
 
-#: Template to use for namespacing a plugin item in the menu bar
-menu_item_template = '{}: {}'
+@lru_cache  # only call once
+def _initialize_plugins():
+    settings = get_settings().plugins
+
+    _npe2pm = _PluginManager.instance()
+    for p in settings.disabled_plugins_npe2:
+        _npe2pm.disable(p)
+    _npe2pm.discover()
+    _npe2pm.events.enablement_changed.connect(
+        _npe2._on_plugin_enablement_change
+    )
+
+    # this is a workaround for the fact that briefcase does not seem to include
+    # napari's entry_points.txt in the bundled app, so the builtin plugins
+    # don't get detected.  So we just register it manually.  This could
+    # potentially be removed when we move to a different bundle strategy
+    if 'napari' not in _npe2pm._manifests:
+        mf_file = Path(__file__).parent.parent / 'builtins.yaml'
+        mf = PluginManifest.from_file(mf_file)
+        mf.package_metadata = PackageMetadata.for_package('napari')
+        _npe2pm.register(mf)
+
+    # Disable plugins listed as disabled in settings, or detected in npe2
+    _from_npe2 = {m.package_metadata.name for m in _npe2pm.iter_manifests()}
+    _from_npe2.add('napari')
+    plugin_manager._skip_packages = _from_npe2
+
+    settings.disabled_plugins = settings.disabled_plugins.difference(
```

Comment:
@nclack, this seems to be working now.  before I push on, would be good to get your thoughts.  this adds a new `disabled_plugins_npe2` field to the settings, and will basically just always remove a field from `disabled_plugins` if it is detected to be an npe2 plugin on this line here. (but will not modify anything it finds in `settings.disabled_plugins_npe2`).  so basically, the first time a plugin transitions to npe2, it will be re-enabled until they disable it again as an npe2 plugin

Reply:
This works but is there a way to avoid having two `disabled_plugins` fields? It looks like you needed to keep two separate lists for the two plugin engines here. Is that right?

Reply:
yeah, that's what I ended up using.  If you can think of a way to keep a single field, that'd be great! :) but it wasn't clear to me how we would determine whether something we found in that one list was an `npe2` plugin that someone had _intended_ to disable, or an `npe2` plugin that had ended up in that list as a side effect of not wanting the npe1 plugin manager to pay attention to it

Reply:
(I guess we could encode it in the string itself (`napari-svg[npe2]` ... but it feels like we'd need to deal with that in a lot of places... forever :joy:)

Reply:
@nclack  have another look now.  After our chat, seemed like the main way to accomplish this was to introduce the concept of migrations (in this case, a migration would just remove any npe2 plugins from the set of `disabled_plugins`).  Would appreciate a somewhat close look here to make sure I'm on the right track, and happy to zoom about it again